File: java-library/src/main/java/com/gblninja/tag/type/GblProg.java
package com.gblninja.tag.type;


import com.gblninja.tag.GblType;
import com.gblninja.tag.Tag;
import com.gblninja.tag.TagHeader;
import com.gblninja.tag.TagWithHeader;

public class GblProg implements Tag, TagWithHeader {
    private final TagHeader tagHeader;
    private final GblType tagType;
    private final long flashStartAddress;
    private final byte[] data;
    private final byte[] tagData;

    public GblProg(TagHeader tagHeader, GblType tagType, long flashStartAddress,
                   byte[] data, byte[] tagData) {
        this.tagHeader = tagHeader;
        this.tagType = tagType;
        this.flashStartAddress = flashStartAddress;
        this.data = data;
        this.tagData = tagData;
    }

    @Override
    public GblType getTagType() {
        return tagType;
    }

    @Override
    public Tag copy() {
        return new GblProg(tagHeader, tagType, flashStartAddress, data, new byte[0]);
    }

    @Override
    public TagHeader getTagHeader() {
        return tagHeader;
    }

    @Override
    public byte[] getTagData() {
        return tagData;
    }

    public long getFlashStartAddress() {
        return flashStartAddress;
    }

    public byte[] getData() {
        return data;
    }
}

File: gbl/src/main/java/com/onix/gbl/parser/type/GblProg.kt
package parser.data.type

import parser.data.tag.TagHeader
import parser.data.tag.TagInterface

data class GblProg(
    override val tagHeader: TagHeader,
    override val tagType: GblType,
    val flashStartAddress: UInt,
    val data: ByteArray,
    override val tagData: ByteArray
): TagInterface  {
    override fun copy(): TagInterface {
        return GblProg(
            tagHeader = tagHeader,
            tagType = tagType,
            flashStartAddress = flashStartAddress,
            data = data,
            tagData = arrayOf<Byte>().toByteArray()
        )
    }
}
